### PR TITLE
Adding Span IndexOf that takes an index and count

### PR DESCRIPTION
**Pull request by ahsonkhan**

### CODE REVIEW

**Comment by ahsonkhan** [comment]
cc @KrzysztofCwalina, @davidfowl

File: src/System.Memory/ref/System.Memory.cs [hunk 1]
```diff
@@ -87,6 +87,7 @@ public static class SpanExtensions
     {
          public static int IndexOf<T>(this Span<T> span, T value) where T:struct, IEquatable<T> { throw null; }
```
Comment (jkotas) [inline src/System.Memory/ref/System.Memory.cs]: Where is the approved API proposal for this? What about the other overloads?

File: src/System.Memory/src/System/SpanHelpers.byte.cs [hunk 2]
```diff
@@ -95,6 +95,69 @@ public static int IndexOf(ref byte searchSpace, byte value, int length)
             return -1;
         }
 
+        public static int IndexOf(ref byte searchSpace, byte value, int startIndex, int count, int spanLength)
```
Comment (jkotas) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: I do not think it makes sense to replicate the worker method for this - you should use the existing `public static int IndexOf(ref byte searchSpace, byte value, int length)` method.

File: src/System.Memory/ref/System.Memory.cs (hunk 1 above)
Comment (jkotas) [inline src/System.Memory/ref/System.Memory.cs]: If this is just a temporary hack until the Span<T> codegen is fixed, you should keep it in corefxlab.

Comment (KrzysztofCwalina) [inline src/System.Memory/ref/System.Memory.cs]: Yeah, I agree corfxlab might be a better place for it. corfx APIs should represent what we are confident in shipping.

Comment (davidfowl) [inline src/System.Memory/ref/System.Memory.cs]: That's fine, will slicing + indexof ever be as fast as passing an index and count? I have some doubts about that.

```C#
while (span.Length > 0)
{
      int index = span.IndexOf((byte)'\n');
      if (index == -1) break;
      var line = span.Slice(0, index);
      span = span.Slice(index + 1);
}
```

VS

```C#
int index = 0;
while (index < span.Length)
{
      int lineIndex = span.IndexOf((byte)'\n', index, span.Length - index);
      if (lineIndex == -1) break;
      var line = span.Slice(index, lineIndex - index + 1);
      index += length.Length
}
```

Will that ever be within the ballpark?

Comment (jkotas) [inline src/System.Memory/ref/System.Memory.cs]: Yes, it should be pretty equivalent - once the JIT team implements the planned struct enregistration that handles Span well.

Comment (davidfowl) [inline src/System.Memory/ref/System.Memory.cs]: Wow, I look forward to seeing this. @ahsonkhan Can you make sure we have a benchmark for this scenario? It's pretty fundamental.

**Comment by ahsonkhan** [comment]
This PR can be closed. Opened one in corefxlab - https://github.com/dotnet/corefxlab/pull/1278

File: src/System.Memory/ref/System.Memory.cs (hunk 1 above)
Comment (ahsonkhan) [inline src/System.Memory/ref/System.Memory.cs]: I have created it in the PR in corefxlab:
```C#
result = bytes.IndexOf(lookupVal, startIndex, count);
```
VS.
```C#
result = bytes.Slice(startIndex, count).IndexOf(lookupVal);
```

**Pull request closed without merging**